### ISSUE TITLE
Storage.put, Storage.get with protected images give 403 error out of the box on localhost

### ISSUE THREAD

**Issue by mayteio** [aws-amplify/amplify-cli#3497]
**Describe the bug**
**When developing locally**, `Storage.put` throws a 403 error out of the box with minimal setup from auth, api and storage categories. If you set the bucket permissions to public read/write access, Storage.put works as expected (though I shouldn't have to do this and is unsafe). When trying to retrieve that same file, even with public read/write access, it returns 403 and I am unable to load the image.

Minimal reproduction repository: https://bitbucket.org/mayteio/amplify-storage-issue/src/master/

My current solution is to use `amplify mock storage` - do I always have to do this?

**Amplify CLI Version**
@aws-amplify/cli 4.13.4

**To Reproduce**
I've created a minimal reproduction of the issue:

```bash
git clone git@bitbucket.org:mayteio/amplify-storage-issue.git
cd amplify-storage-issue
yarn
amplify init
amplify push
yarn start
```

1. sign up for an account in the app when it opens in the browser
2. after confirmation, attempt to upload an image using `Storage.put` and you get 403 instantly. ** That's the first issue **.

Log into your console, visit your s3 bucket permissions -> ACL -> public read/write -> Save. 

![Screen Shot 2020-02-24 at 10 36 37 am](https://user-images.githubusercontent.com/43975092/75122434-8fecf680-56f1-11ea-96d5-dbb960a7d412.png)

Now upload goes smoothly. Give your image a title and hit save. Success, your console should show something like this:

```
{
  "data": {
    "createImage": {
      "id": "089bb095-6ace-447b-b86e-2e1a022b36cf",
      "title": "Title",
      "file": [
        {
          "key": "2.032109714444997Screen Shot 2020-02-24 at 9.48.35 am.png",
          "identityId": "ap-southeast-2:2201d300-5134-4538-83ec-911d281d3eab"
        }
      ],
      "owner": "harley"
    }
  }
}
```

In, `ImagesList.js`, I use the S3Image component provided to display it. Now, **despite setting the ACL to public read/write** I get a 403 Error. Same issue if I write my own implementation.

```
const CustomS3Image = ({ file }) => {
  const [image, setImage] = useState();
  useEffect(() => {
    Storage.get(file.key, {
      level: 'protected',
      identityId: file.identityId,
    })
      .then(setImage)
      .catch(err => console.log(err));
  }, [file]);

  return image && <img src={image} alt="" />;
};
```

**Expected behavior**
`Storage.put` should work out of the box on localhost for development. I have to deploy it to s3 bucket hosting to get it to work. This is a terrible feedback loop.

**Screenshots**
Localhost failing:
![Screen Shot 2020-02-24 at 11 18 19 am](https://user-images.githubusercontent.com/43975092/75122982-8797ba00-56f7-11ea-9777-644968118ad3.png)

Working on deployed s3 bucket:
![Screen Shot 2020-02-24 at 11 19 21 am](https://user-images.githubusercontent.com/43975092/75122983-8e263180-56f7-11ea-86b9-3905828bc1c9.png)

**Desktop (please complete the following information):**
- macOS 10.15.3
- node 13.8.0

**Additional context**
Given Amplify's promise is easy development, I find it hilarious I have to jump through so many hoops to get something so fundamental working. Pretty frustrating. Or that I have to store `identityId` in dyanmodb to retrieve protected images... but that's another story.

Questions for the amplify team this leaves me with
- Is it safe to store the user's `identityId next to the S3 key, as we need it for retrieval?
- Is there any additional setup after running `amplify add storage` that I have to complete in the console or elsewhere? What am I missing here?

**Related issues**
- https://github.com/aws-amplify/amplify-cli/issues/1847
- https://github.com/aws-amplify/amplify-js/issues/1094

**Comment by nikhname** [comment]
@mayteio Did you enable triggers for Cognito (should be prompted on init with your repro steps)? If so which triggers did you enable?

**Comment by mayteio** [comment]
Yes, just the post confirmation trigger with addUserToGroup lambda.

**Comment by nikhname** [comment]
@mayteio I was not able to reproduce this using your project, Storage.put and Storage.get worked out of the box for me. Could you please review the IAM permissions for your account?
https://aws.amazon.com/iam/features/manage-permissions/

**Comment by mayteio** [comment]
Thanks for getting back to me @nikhname, was that on localhost or when deployed? Because it works fine when deployed, just not testing locally.

**Comment by amuresia** [comment]
I can reproduce this even if I add hosting and deploy the react app ...
I am using `@auth(rules: [{ allow: groups, groups: ["Admin"], operations: [create update delete] }])` and I get a 403 on upload to S3.
What's even more concerning is that if the user that I am logged in with is not part of the `Admin` group then the mutation fails **but the upload is successful** ... yikes.

**Comment by akshbhu** [comment]
Hi @mayteio 
I tried with your above instructions to reproduce the error on localhost but its working fine for me.
Both Storage.put and Storage.get are working fine on localhost, I can see the images also as you are displaying them side by side with the image name also.
If you are still stuck ,pls let me know.

**Comment by amuresia** [comment]
Yes, I was able to reproduce this today. It's still an issue. I have to disable auth on my graphql model for the Storage.put to be successful.

**Comment by mayteio** [comment]
Came across the same issue again on another project. This is definitely reproducible across projects.

**Comment by ajpaulingalls** [comment]
I am getting a 403 when using Storage.put as well, which started when I upgraded to latest versions of amplify-js...

**Comment by lukepushlabs** [comment]
Also having this issue, happened when upgrading recently.

**Comment by akshbhu** [comment]
Hi @lukepushlabs @mayteio  @ajpaulingalls 

Can you tell me the steps like which to which versions of amplify-cli and amplify-js you upgraded to so that I can reproduce the above error.

**Comment by ajpaulingalls** [comment]
I dropped my details into this bug in amplify-js:

https://github.com/aws-amplify/amplify-js/issues/5273

I'm guessing it has more to do with their changes than amplify-cli...

**Comment by wadamomo** [comment]
I also randomly started getting 403 on Storage.put for images/blobs a few days ago running aws-amplify@3.0.5-unstable.4 

I could still upload a text string using Storage.put and it would be going into the right bucket but failed for images. 

Just downgraded and it is now working fine again.

**Comment by jingxizhang** [comment]
I am having the same 403 forbidden issue as reported by @amuresia. My graphql schema models all have static Cognito group authorization (@auth(rules: [{ allow: groups, groups: ["Admin", "Engineer", ...]})). The S3 storage was created by using Amplify Cli "amplify add storage". The access restriction is by individual groups and a lambda trigger is also added:

? Restrict access by? Individual Groups
? Select groups: Admin, FieldEngineer, Manufactory, Engineer
? What kind of access do you want for Admin users? create/update, read, delete
? What kind of access do you want for FieldEngineer users? create/update, read
? What kind of access do you want for Manufactory users? create/update, read
? What kind of access do you want for Engineer users? create/update, read, delete
? Do you want to add a Lambda Trigger for your S3 Bucket? Yes
? Select from the following options Create a new function
Successfully added resource S3Triggerxxxxx locally

I loaded some images into the S3 bucket. The React client calls Storage.get(key) and it returns the signed URL without error (return code 200). However, the signed URL is set to the src of <img> HTML tag and it receives 403 forbidden error. 
I have tried to change the S3 restrict access by "Auth users only", the problem is the same.
I also changed the React code by directly using S3Image from 'aws-amplify-react', it has the same 403 forbidden error.
My amplify cli version is 4.18.0.

**Comment by ahansson89** [comment]
For what it is worth, I am facing this issue too.

**Comment by jsgphd** [comment]
Version 3.0.8 is working for me.

**Comment by spyboost** [comment]
> Version 3.0.8 is working for me.

3.0.8 Doesn't work for me. Getting 403.

**Comment by jingxizhang** [comment]
I found the problem: the Storage.get() inserts "/public/" in between S3 bucket endpoint and the S3 key in the returned pre-signed url. For an image file when using this pre-signed url as the image src you will get the 403 forbidden error (it should return 404 file not found error, instead).
I tried to work around the problem by removing the "/public/" from the presigned url. However, I got an error "SignatureDoesNotMatch" return because it is a signed url I cannot change it.
If I create a "public" folder in my S3 bucket and move my image files into the "public" folder then it works. 
Is there anyway to configure the Store sub-module not insert the "/public/" in the returned url?

**Comment by spyboost** [comment]
@jingxizhang - I think this is how it works by design. Those access levels are documented here: https://docs.amplify.aws/lib/storage/configureaccess/q/platform/js
So everything goes to /public by default.
I don't know if it's possible to configure it differently.

**Comment by jingxizhang** [comment]
@spyboost Thank you for the explaining.
I tried to change the Storage level using "public", "private" or empty string, I got different results.
The code snippet: 
```
import AmplifyConfig from '../aws-exports'
const {
    aws_user_files_s3_bucket, 
    aws_user_files_s3_bucket_region, 
    aws_cognito_identity_pool_id
} = AmplifyConfig

Storage.configure({
    bucket: aws_user_files_s3_bucket,
    level: "private",
    region: aws_user_files_s3_bucket_region,
    identityPoolId: aws_cognito_identity_pool_id
})

```
When I use "public" or empty string for the level the resulting url is s3Endpoint + "/public" + s3Key, as expected.
However, when I use "private" for the level the resulting url is s3Endpoint + "/private/us-west-2%3Af07b89c6-280c-4a05-9c4b-f76ba6746eb0" + s3Key. Not sure what is the numbers following "private/us-west-2" is. I could not find the document for it.
Can you help me for that?

**Comment by spyboost** [comment]
@jingxizhang - unfortunately, I'm not an expert and learning this new stuff myself. My understanding is that this looks like the user id from Cognito.

**Comment by mayteio** [comment]
That’s the identityId of the user who uploaded the file. Also by design. When you retrieve it you need to pass in the identityId. I store this against my S3Object in dynamodb as it’s the only way to reliably retrieve it. No idea if that’s a security flaw or not.

**Comment by spyboost** [comment]
@jingxizhang - also the whole configuration can be simplified to just calling 
```
Amplify.configure(awsExports);
```
From the docs:
> If you use aws-exports.js file, Storage is already configured when you call Amplify.configure(awsconfig).

**Comment by spyboost** [comment]
@mayteio - so the upload is working for you with 3.8.0? You're lucky. In my case I can't even upload to /protected :(

**Comment by mayteio** [comment]
Yes, but only when it is hosted. Not from localhost. Have you tried deploying it and testing? Or try using amplify mock to save yourself some time.

**Comment by spyboost** [comment]
@mayteio nope local doesn't work for me even with mock. I think there's a separate github issue for it as it's trying to send to https which fails locally, although I'm not even sure right now as I got lost tracking this problem from many different angles.

**Comment by jingxizhang** [comment]
@mayteio Thank you.
How do I get the identityId of the user? I checked the Cognito user pool, the logged in user's sub is different from the number in the url.
I would like all the users are able to read the image files in the S3, which level I should user? According to AWS amplify document. The "protected" level should allow all users to read the file but only the owner can write to the file. However, I checked with "protected" level, Storage.get will automatically insert a different identity number in the url. However, can the user to access the file shared (read only)?

**Comment by mayteio** [comment]
I believe from Auth.getCurrentCredentials. It’s in. The id token from cognito. Yes if any user has the identityId of the uploader they can see protected files.

**Comment by jingxizhang** [comment]
How can user B get the user A's identityId who is the uploader?

**Comment by jingxizhang** [comment]
Even if user B get the uploader's identityId, how can Storage to set to user uploader's identityId in the generated pre-signed url?

**Comment by spyboost** [comment]
@jingxizhang - this appears to be a very huge torture. 
https://github.com/aws-amplify/amplify-js/issues/2203

**Comment by mayteio** [comment]
As mentioned, you can store it against an object at upload time to dynamodb. 

‘’’
type S3Object {
  key: String
  level: String
  identityId: String
}
‘’’

I agree it’s torture but it’s the way they’ve built it.

**Comment by jingxizhang** [comment]
@mayteio - thanks!
In my case, I have many images uploaded from other project (the IoT Core project) which security model is not through Cognito but using X.509 certificate (so there is no uploader's identityId). I am not sure how to share those images.

**Comment by mayteio** [comment]
@akshbhu what do you need here to move this one forward? There is a “not reproducible” tag however plenty of people have reproduced it with instructions.

**Comment by mayteio** [comment]
@jingxizhang no problem. I am not sure either without having my laptop. Maybe open another issue as that is unrelated to this one.

**Comment by spyboost** [comment]
@mayteio - there might be a way to not store this in dynamodb, depending on how your graphql entities are stored/retrieved, of course. For example, in a hypothetical scenario, if you have a new cool dating app and return a list of user profiles, each item UserProfile already has the id in response which then is used to query the data from Storage. Just wanted to double check I'm correct in my assumption.

**Comment by mayteio** [comment]
It would have their sub (cognito pool Id) but not their identityId from the identity pool. This is why you have to store it as it’s generated during authentication not account creation. You could store it once against the user in attributes and update it every time they log in. My solution above is the cleanest and easiest I have found though.

**Comment by spyboost** [comment]
@mayteio thanks for clarification. At this point I'm confused even more. I'm looking at the response from currentAuthenticatedUser() and I don't understand what property is the identityId. There's no such field. And there are a lot of other properties that look like a potential id, but none of the values are being used to send data for Storage.Put after looking at the request in chrome console.

**Comment by mayteio** [comment]
https://github.com/mayteio/ra-aws-amplify/blob/112dd87fc2b2187266deb01b2b1b73e13ee1e6d0/src/AmplifyAuthProvider/AmplifyAuthProvider.tsx#L87

`Auth.currentCredentials().then(({identityId}) => console.log(identityId))`

Please open another issue if you still can’t, trying to keep this one on track.

**Comment by danrivett** [comment]
Our project has just run into this issue too. We're attempting to downgrade to an older version to see if that helps, but I would love an update from @akshbhu or someone from AWS to at least find out if this is being investigated yet, or if this is still in the backlog to look at.

Being transparent, this is quite a frustrating bug to run into, affecting our velocity, because we're not doing anything weird or complicated.

**Comment by akshbhu** [comment]
Hi @danrivett 

We are looking in this issue.I tried reproducing it earlier but no luck. But it seems it occurring frequently.
Can you provide the steps i.e like categories you added so that I can reproduce it ?

**Comment by danrivett** [comment]
@akshbhu Yeah a developer on my team reported this to me today and couldn't figure out what's going on with receiving a 403.

I'm going to attempt to recreate it myself later today and will report back with repro steps hopefully.

**Comment by mtakac** [comment]
Hi, same issue for me. I tried `@aws-amplify/storage` `v3.1.7`, `v3.1.8` and `v3.1.9`. 

`amplify status`:

**Comment by mtakac** [comment]
Hi, same issue for me. I tried `@aws-amplify/storage` `v3.1.7`, `v3.1.8` and `v3.1.9`. 

**storage was added with these settings:**

```
? Please select from one of the below mentioned services: Content (Images, audio, video, etc.)
? Restrict access by? Auth/Guest Users
? Who should have access: Auth and guest users
? What kind of access do you want for Authenticated users? create/update, read, delete
? What kind of access do you want for Guest users? create/update, read, delete
? Do you want to add a Lambda Trigger for your S3 Bucket? No
```

**amplify status**:

```
| Category | Resource name             | Operation | Provider plugin   |
| -------- | ------------------------- | --------- | ----------------- |
| Function |  | No Change | awscloudformation |
| Function |  | No Change | awscloudformation |
| Auth     |  | No Change | awscloudformation |
| Auth     |  | No Change | awscloudformation |
| Api      |  | No Change | awscloudformation |
| Api      |  | No Change | awscloudformation |
| Storage  |  | No Change | awscloudformation |
```

**debug log:**

```
[DEBUG] 34:14.647 Credentials - credentials not changed and not expired, directly return
[DEBUG] 34:14.651 AWSS3Provider - set credentials for storage Object {
  "accessKeyId": "XXX",
  "authenticated": true,
  "identityId": "XXX",
  "secretAccessKey": "XXX",
  "sessionToken": "XXX",
}
[DEBUG] 34:14.660 AWSS3Provider - put XXX.jpg to public/XXX.jpg
[DEBUG] 34:15.779 axios-http-handler Event {
  "isTrusted": false,
  "lengthComputable": true,
  "loaded": 32768,
  "total": 1303178,
}
[DEBUG] 34:15.796 axios-http-handler Event {
  "isTrusted": false,
  "lengthComputable": true,
  "loaded": 196608,
  "total": 1303178,
}
[DEBUG] 34:20.797 axios-http-handler Event {
  "isTrusted": false,
  "lengthComputable": true,
  "loaded": 393216,
  "total": 1303178,
}
[DEBUG] 34:23.146 axios-http-handler Event {
  "isTrusted": false,
  "lengthComputable": true,
  "loaded": 1179648,
  "total": 1303178,
}
[DEBUG] 34:23.404 axios-http-handler Event {
  "isTrusted": false,
  "lengthComputable": true,
  "loaded": 1303178,
  "total": 1303178,
}

[ERROR] 34:23.880 axios-http-handler, [Error: Request failed with status code 403]
- node_modules/expo/build/environment/muteWarnings.fx.js:27:24 in error
* [native code]:null in error
- node_modules/@aws-amplify/core/lib-esm/Logger/ConsoleLogger.js:96:8 in prototype._log
* http://127.0.0.1:19001/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&minify=false&hot=false:124334:22 in error
- node_modules/@aws-amplify/storage/lib-esm/providers/axios-http-handler.js:90:23 in axios.request.then._catch$argument_0
- node_modules/promise/setimmediate/core.js:37:14 in tryCallOne
- node_modules/promise/setimmediate/core.js:123:25 in setImmediate$argument_0
- node_modules/react-native/Libraries/Core/Timers/JSTimers.js:146:14 in _callTimer
- node_modules/react-native/Libraries/Core/Timers/JSTimers.js:194:17 in _callImmediatesPass
- node_modules/react-native/Libraries/Core/Timers/JSTimers.js:458:30 in callImmediates
* [native code]:null in callImmediates
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:407:6 in __callImmediates
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:143:6 in __guard$argument_0
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:384:10 in __guard
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:142:17 in __guard$argument_0
* [native code]:null in flushedQueue
* [native code]:null in callFunctionReturnFlushedQueue
```

**code in React Native app:**

```
import { Storage } from '@aws-amplify/storage';

await Storage.put('XXX', blob, {
    contentType: blob.type,
});
```

**Comment by brene** [comment]
Hi folks - diving into this more, it seems like it's a JS library issue. I'm transferring this issue to the JS repo. cc @sammartinez
